File: frontend/js/government.js
/**
 * AEGIS — Government Intelligence Command Center
 * Frontend logic: auth, dashboard, feed, alerts, charts, detail modal
 */
(function () {
    'use strict';

    const API = 'http://127.0.0.1:8001';
    let authToken = localStorage.getItem('aegis_token') || null;
    let currentAdvisoryId = null;
    let refreshTimer = null;

    // ── Charts ──
    let chartAir, chartHeat, chartComposite;

    // ══════════════════════════════════════════════════════
    // AUTH
    // ══════════════════════════════════════════════════════

    function showLogin() {
        console.log('Showing login screen');
        const loginScreen = document.getElementById('loginScreen');
        const mainApp = document.getElementById('mainApp');
        console.log('Login screen element:', loginScreen);
        console.log('Main app element:', mainApp);
        loginScreen.style.display = 'flex';
        mainApp.style.display = 'none';
        console.log('Login screen display:', loginScreen.style.display);
        console.log('Main app display:', mainApp.style.display);
    }

    function showApp() {
        console.log('Showing main app');
        const loginScreen = document.getElementById('loginScreen');
        const mainApp = document.getElementById('mainApp');
        console.log('Login screen element:', loginScreen);
        console.log('Main app element:', mainApp);
        loginScreen.style.display = 'none';
        mainApp.style.display = 'block';
        console.log('Login screen display:', loginScreen.style.display);
        console.log('Main app display:', mainApp.style.display);
        initCharts();
        loadDashboard();
        startAutoRefresh();
    }

    async function login() {
        console.log('Login function called');
        const user = document.getElementById('loginUser').value.trim();
        const pass = document.getElementById('loginPass').value;
        const errEl = document.getElementById('loginError');
        console.log('Username:', user);
        console.log('Password length:', pass.length);
        errEl.style.display = 'none';

        try {
            const resp = await fetch(`${API}/api/gov/login`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ username: user, password: pass }),
            });
            console.log('Login response status:', resp.status);
            if (!resp.ok) throw new Error('Auth failed');
            const data = await resp.json();
            console.log('Login response data:', data);
            authToken = data.access_token;
            localStorage.setItem('aegis_token', authToken);
            console.log('Token stored, showing app');
            showApp();
        } catch (error) {
            console.error('Login error:', error);
            errEl.style.display = 'block';
        }
    }

    function logout() {
        authToken = null;
        localStorage.removeItem('aegis_token');
        if (refreshTimer) clearInterval(refreshTimer);
        showLogin();
    }

    function authHeaders() {
        return { Authorization: `Bearer ${authToken}` };
    }

    async function apiFetch(url) {
        console.log(`Making API call to: ${API}${url}`);
        try {
            const resp = await fetch(`${API}${url}`, { headers: authHeaders() });
            console.log(`API response status: ${resp.status}`);
            if (resp.status === 401) { 
                console.log('Unauthorized access, logging out');
                logout(); 
                throw new Error('Unauthorized'); 
            }
            if (!resp.ok) {
                console.error(`API call failed with status: ${resp.status}`);
                throw new Error(`HTTP error! status: ${resp.status}`);
            }
            const data = await resp.json();
            console.log(`API response data:`, data);
            return data;
        } catch (error) {
            console.error(`API fetch error for ${url}:`, error);
            throw error;
        }
    }

    // ══════════════════════════════════════════════════════
    // DASHBOARD
    // ══════════════════════════════════════════════════════

    async function loadDashboard() {
        console.log('Loading dashboard...');
        try {
            const data = await apiFetch('/api/gov/dashboard');
            console.log('Dashboard data:', data);
            updateRiskCard('air', data.air);
            updateRiskCard('heat', data.heat);
            updateRiskCard('composite', data.composite);

            document.getElementById('alertCount').textContent = data.active_alerts;
            if (data.last_updated) {
                const d = new Date(data.last_updated);
                document.getElementById('lastUpdated').textContent = d.toLocaleTimeString('en-IN', { hour: '2-digit', minute: '2-digit' });
            }
        } catch (e) {
            console.warn('Dashboard load error:', e);
        }

        loadQuickFeed();
        loadAlerts();
        loadTrends();
        loadHealth();
    }

    function updateRiskCard(type, d) {
        console.log(`Updating risk card for ${type}:`, d);
        if (!d) {
            console.log(`No data for ${type} risk card`);
            return;
        }
        const score = d.score != null ? d.score.toFixed(1) : '--';
        const scoreElement = document.getElementById(`${type}Score`);
        console.log(`${type}Score element:`, scoreElement);
        if (scoreElement) {
            scoreElement.innerHTML = `${score}<span class="risk-card-score-unit">/100</span>`;
        }

        // Status badge
        const statusEl = document.getElementById(`${type}Status`);
        console.log(`${type}Status element:`, statusEl);
        if (statusEl) {
            statusEl.textContent = d.status || '--';
            statusEl.className = 'risk-card-status status-' + (d.status || 'normal').toLowerCase();
        }

        // Value
        if (type === 'air') {
            const airValueElement = document.getElementById('airValue');
            console.log('airValue element:', airValueElement);
            if (airValueElement) {
                airValueElement.textContent = `PM2.5: ${d.value != null ? d.value.toFixed(1) : '--'} ${d.unit || 'µg/m³'}`;
            }
        } else if (type === 'heat') {
            const heatValueElement = document.getElementById('heatValue');
            console.log('heatValue element:', heatValueElement);
            if (heatValueElement) {
                heatValueElement.textContent = `Heat Index: ${d.value != null ? d.value.toFixed(1) : '--'} ${d.unit || '°C'}`;
            }
        } else {
            const compositeValueElement = document.getElementById('compositeValue');
            console.log('compositeValue element:', compositeValueElement);
            if (compositeValueElement) {
                compositeValueElement.textContent = `Amplification: ${d.amplification ? d.amplification.toFixed(3) : '--'}`;
            }
        }

        // Confidence
        const confidenceElement = document.getElementById(`${type}Confidence`);
        console.log(`${type}Confidence element:`, confidenceElement);
        if (confidenceElement) {
            confidenceElement.textContent = `Confidence: ${d.confidence != null ? d.confidence.toFixed(0) : '--'}%`;
        }

        // Trend
        const trendEl = document.getElementById(`${type}Trend`);
        console.log(`${type}Trend element:`, trendEl);
        if (trendEl) {
            const trendMap = { up: '↑ Rising', down: '↓ Falling', stable: '— Stable' };
            trendEl.textContent = trendMap[d.trend] || '— Stable';
            trendEl.className = `risk-card-trend trend-${d.trend || 'stable'}`;
        }
    }

    // ══════════════════════════════════════════════════════
    // INTELLIGENCE FEED
    // ══════════════════════════════════════════════════════

    async function loadQuickFeed() {
        try {
            const data = await apiFetch('/api/gov/feed?limit=8');
            renderFeed('quickFeed', data);
        } catch (e) { console.warn('Feed error:', e); }
    }

    async function loadFullFeed() {
        try {
            const data = await apiFetch('/api/gov/feed?limit=50');
            renderFeed('fullFeed', data);
        } catch (e) { console.warn('Feed error:', e); }
    }

    function renderFeed(containerId, entries) {
        const container = document.getElementById(containerId);
        if (!entries.length) {
            container.innerHTML = '<div class="empty-state">No intelligence entries</div>';
            return;
        }
        container.innerHTML = entries.map(e => {
            const time = e.timestamp ? new Date(e.timestamp).toLocaleTimeString('en-IN', { hour: '2-digit', minute: '2-digit' }) : '--:--';
            const severityClass = `feed-severity-${e.severity || 'info'}`;
            return `
                <div class="feed-entry" onclick="this.classList.toggle('expanded')">
                    <div class="feed-entry-header">
                        <div class="feed-entry-title"><span class="feed-severity-badge ${severityClass}"></span>${e.title}</div>
                        <div class="feed-entry-time">${time}</div>
                    </div>
                    <div class="feed-entry-content">${e.content}</div>
                </div>`;
        }).join('');
    }

    // ══════════════════════════════════════════════════════
    // ALERTS
    // ══════════════════════════════════════════════════════

    async function loadAlerts() {
        try {
            const data = await apiFetch('/api/gov/alerts');
            renderAlerts('quickAlerts', data.active, true);
            renderAlerts('activeAlerts', data.active, true);
            renderAlerts('pastAlerts', data.past, false);
        } catch (e) { console.warn('Alerts error:', e); }
    }

    function renderAlerts(containerId, alerts, showActions) {
        const container = document.getElementById(containerId);
        if (!container) return;
        if (!alerts || !alerts.length) {
            container.innerHTML = '<div class="empty-state">No alerts</div>';
            return;
        }
        container.innerHTML = alerts.map(a => {
            const time = a.timestamp ? new Date(a.timestamp).toLocaleTimeString('en-IN', { hour: '2-digit', minute: '2-digit', hour12: true }) : '';
            const sevClass = `alert-item-severity-${(a.severity || 'normal').toLowerCase()}`;
            return `
                <div class="alert-item ${sevClass}">
                    <div class="alert-item-body">
                        <div class="alert-item-message">${a.message}</div>
                        <div class="alert-item-meta">${a.severity} · ${a.type.replace(/_/g, ' ')} · ${time}</div>
                    </div>
                    ${showActions ? `
                        <div class="alert-item-actions">
                            <button class="aegis-btn aegis-btn-sm" onclick="acknowledgeAlert(${a.id})">ACK</button>
                        </div>
                    ` : ''}
                </div>`;
        }).join('');
    }

    window.acknowledgeAlert = async function (id) {
        try {
            await fetch(`${API}/api/gov/alerts/${id}/acknowledge`, {
                method: 'POST',
                headers: authHeaders(),
            });
            loadAlerts();
            loadDashboard();
        } catch (e) { console.warn('Ack error:', e); }
    };

    // ══════════════════════════════════════════════════════
    // ADVISORY
    // ══════════════════════════════════════════════════════

    async function generateAdvisory() {
        try {
            const resp = await fetch(`${API}/api/gov/advisory/generate`, {
                method: 'POST',
                headers: authHeaders(),
            });
            const data = await resp.json();
            currentAdvisoryId = data.id;
            document.getElementById('advisoryDraftText').textContent = data.message;
            document.getElementById('advisoryDraft').style.display = 'block';
        } catch (e) { console.warn('Advisory gen error:', e); }
    }

    async function approveAdvisory() {
        if (!currentAdvisoryId) return;
        try {
            await fetch(`${API}/api/gov/advisory/${currentAdvisoryId}/approve`, {
                method: 'POST',
                headers: authHeaders(),
            });
            document.getElementById('advisoryDraft').style.display = 'none';
            currentAdvisoryId = null;
            loadDashboard();
        } catch (e) { console.warn('Advisory approve error:', e); }
    }

    // ══════════════════════════════════════════════════════
    // TRENDS / CHARTS
    // ══════════════════════════════════════════════════════

    function initCharts() {
        console.log('Initializing charts...');
        try {
            chartAir = new AegisChart('chartAir', { yLabel: 'PM2.5 (µg/m³)', lineColor: '#A855F7' });
            console.log('Chart Air initialized:', chartAir);
            chartHeat = new AegisChart('chartHeat', { yLabel: 'Heat Index (°C)', lineColor: '#C084FC' });
            console.log('Chart Heat initialized:', chartHeat);
            chartComposite = new AegisChart('chartComposite', { yLabel: 'Risk Score', yMin: 0, yMax: 100, lineColor: '#7C3AED' });
            console.log('Chart Composite initialized:', chartComposite);
        } catch (error) {
            console.error('Error initializing charts:', error);
        }
    }

    async function loadTrends() {
        try {
            const data = await apiFetch('/api/gov/trends?hours=24');
            if (chartAir && data.pm25_values) chartAir.setData(data.labels, data.pm25_values);
            if (chartHeat && data.heat_index_values) chartHeat.setData(data.labels, data.heat_index_values);
            if (chartComposite && data.composite_scores) chartComposite.setData(data.labels, data.composite_scores);
        } catch (e) { console.warn('Trends error:', e); }
    }

    // ══════════════════════════════════════════════════════
    // SYSTEM HEALTH
    // ══════════════════════════════════════════════════════

    async function loadHealth() {
        // Engine agent statuses
        try {
            const data = await apiFetch('/api/gov/system-health');
            setHealth('healthObserver', data.observer_agent);
            setHealth('healthDetection', data.detection_engine);
            setHealth('healthInvestigation', data.investigation_engine);
            setHealth('healthExplanation', data.explanation_engine);
            setHealth('healthAutonomy', data.autonomy_mode);
        } catch (e) { console.warn('Health error:', e); }

        // Live API connectivity
        try {
            const resp = await fetch(`${API}/system/health`);
            const api = await resp.json();
            setApiHealth('healthOpenAQ', api.openaq);
            setApiHealth('healthOpenWeather', api.openweather);
            setApiHealth('healthOpenRouter', api.openrouter);
        } catch (e) { console.warn('API health error:', e); }
    }

    function setHealth(elId, info) {
        const el = document.getElementById(elId);
        if (!el || !info) return;
        el.textContent = info.status;
        el.className = 'health-item-status' + (info.status === 'Enabled' || info.status === 'Active' || info.status === 'Running' || info.status === 'Operational' || info.status === 'Connected' ? '' : ' inactive');
    }

    function setApiHealth(elId, status) {
        const el = document.getElementById(elId);
        if (!el) return;
        const isOk = status === 'connected';
        el.textContent = isOk ? 'Connected' : (status || 'Unknown');
        el.className = 'health-item-status' + (isOk ? '' : ' inactive');
    }

    // ══════════════════════════════════════════════════════
    // DETAIL MODAL
    // ══════════════════════════════════════════════════════

    async function openDetail(riskType) {
        try {
            const data = await apiFetch(`/api/gov/risk/${riskType}/detail`);
            const modal = document.getElementById('detailModal');
            const titleMap = { air: 'Air Pollution Risk — Detailed Intelligence', heat: 'Heat Stress Risk — Detailed Intelligence', composite: 'Composite Risk — Detailed Intelligence' };
            document.getElementById('modalTitle').textContent = titleMap[riskType] || 'Risk Detail';

            let gridHTML = '';
            gridHTML += detailItem('Risk Score', data.score != null ? data.score.toFixed(1) + '/100' : '--');
            gridHTML += detailItem('Confidence', data.confidence != null ? data.confidence.toFixed(0) + '%' : '--');

            if (riskType === 'air') {
                gridHTML += detailItem('Current PM2.5', data.current_value != null ? data.current_value.toFixed(1) + ' µg/m³' : '--');
                gridHTML += detailItem('Baseline Mean', data.baseline_mean != null ? data.baseline_mean.toFixed(1) + ' µg/m³' : 'Building...');
                gridHTML += detailItem('Deviation (Z-score)', data.zscore != null ? data.zscore.toFixed(3) : '--');
                gridHTML += detailItem('Historical Percentile', data.percentile != null ? data.percentile.toFixed(1) + 'th' : '--');
            } else if (riskType === 'heat') {
                gridHTML += detailItem('Heat Index', data.current_value != null ? data.current_value.toFixed(1) + ' °C' : '--');
                gridHTML += detailItem('Temperature', data.temperature != null ? data.temperature.toFixed(1) + ' °C' : '--');
                gridHTML += detailItem('Humidity', data.humidity != null ? data.humidity.toFixed(0) + '%' : '--');
                gridHTML += detailItem('Baseline Mean', data.baseline_mean != null ? data.baseline_mean.toFixed(1) + ' °C' : 'Building...');
                gridHTML += detailItem('Deviation (Z-score)', data.zscore != null ? data.zscore.toFixed(3) : '--');
                gridHTML += detailItem('Historical Percentile', data.percentile != null ? data.percentile.toFixed(1) + 'th' : '--');
            } else {
                gridHTML += detailItem('Air Contribution', data.air_contribution != null ? data.air_contribution.toFixed(1) : '--');
                gridHTML += detailItem('Heat Contribution', data.heat_contribution != null ? data.heat_contribution.toFixed(1) : '--');
                gridHTML += detailItem('Amplification Factor', data.amplification_factor != null ? data.amplification_factor.toFixed(4) : '--');
                gridHTML += detailItem('Synergistic Boost', data.synergistic_boost != null ? '+' + data.synergistic_boost.toFixed(1) : '--');
                gridHTML += detailItem('Air Risk Score', data.air_score != null ? data.air_score.toFixed(1) + '/100' : '--');
                gridHTML += detailItem('Heat Risk Score', data.heat_score != null ? data.heat_score.toFixed(1) + '/100' : '--');
            }

            document.getElementById('modalGrid').innerHTML = gridHTML;
            document.getElementById('modalSummaryText').textContent = data.intelligence_summary || 'No intelligence summary available — awaiting LLM processing.';

            if (data.anomaly_flags && data.anomaly_flags.length) {
                document.getElementById('modalSummaryText').textContent += '\n\n⚠ Anomaly Flags: ' + data.anomaly_flags.join(', ');
            }

            modal.classList.add('active');
        } catch (e) {
            console.warn('Detail error:', e);
        }
    }

    function detailItem(label, value) {
        return `
            <div class="modal-detail-item">
                <div class="modal-detail-label">${label}</div>
                <div class="modal-detail-value">${value}</div>
            </div>`;
    }

    // ══════════════════════════════════════════════════════
    // TABS
    // ══════════════════════════════════════════════════════

    function switchTab(tabId) {
        document.querySelectorAll('.aegis-tab').forEach(t => t.classList.remove('active'));
        document.querySelectorAll('.tab-content').forEach(t => t.classList.remove('active'));
        document.querySelector(`.aegis-tab[data-tab="${tabId}"]`).classList.add('active');
        document.getElementById(`tab-${tabId}`).classList.add('active');

        if (tabId === 'feed') loadFullFeed();
        if (tabId === 'analytics') {
            loadTrends();
            setTimeout(() => {
                if (chartAir) chartAir._resize();
                if (chartHeat) chartHeat._resize();
                if (chartComposite) chartComposite._resize();
            }, 100);
        }
        if (tabId === 'alerts') loadAlerts();
        if (tabId === 'health') loadHealth();
    }

    // ══════════════════════════════════════════════════════
    // AUTO REFRESH
    // ══════════════════════════════════════════════════════

    function startAutoRefresh() {
        if (refreshTimer) clearInterval(refreshTimer);
        refreshTimer = setInterval(loadDashboard, 60000);
    }

    // ══════════════════════════════════════════════════════
    // TRIGGER CYCLE
    // ══════════════════════════════════════════════════════

    async function triggerCycle() {
        try {
            const btn = document.getElementById('triggerCycleBtn');
            btn.textContent = '⟳ ...';
            btn.disabled = true;
            await fetch(`${API}/api/trigger-cycle`, { method: 'POST' });
            await loadDashboard();
            btn.textContent = '⟳ CYCLE';
            btn.disabled = false;
        } catch (e) {
            console.warn('Cycle trigger error:', e);
            document.getElementById('triggerCycleBtn').textContent = '⟳ CYCLE';
            document.getElementById('triggerCycleBtn').disabled = false;
        }
    }

    // ══════════════════════════════════════════════════════
    // INIT
    // ══════════════════════════════════════════════════════

    document.addEventListener('DOMContentLoaded', () => {
        console.log('DOM loaded, initializing AEGIS...');
        try {
            // Login
            const loginBtn = document.getElementById('loginBtn');
            const loginPass = document.getElementById('loginPass');
            const logoutBtn = document.getElementById('logoutBtn');
            
            if (loginBtn) {
                loginBtn.addEventListener('click', login);
                console.log('Login button event listener added');
            } else {
                console.error('Login button not found!');
            }
            
            if (loginPass) {
                loginPass.addEventListener('keydown', e => { if (e.key === 'Enter') login(); });
                console.log('Login password Enter key event listener added');
            } else {
                console.error('Login password field not found!');
            }
            
            if (logoutBtn) {
                logoutBtn.addEventListener('click', logout);
                console.log('Logout button event listener added');
            } else {
                console.error('Logout button not found!');
            }

            // Tabs
            const tabs = document.querySelectorAll('.aegis-tab');
            console.log(`Found ${tabs.length} tab elements`);
            tabs.forEach(tab => {
                tab.addEventListener('click', () => switchTab(tab.dataset.tab));
                console.log(`Tab listener added for ${tab.dataset.tab}`);
            });

            // Risk cards click → detail modal
            const riskCards = document.querySelectorAll('.risk-card');
            console.log(`Found ${riskCards.length} risk card elements`);
            riskCards.forEach(card => {
                card.addEventListener('click', () => openDetail(card.dataset.risk));
                console.log(`Risk card listener added for ${card.dataset.risk}`);
            });

            // Modal close
            const modalClose = document.getElementById('modalClose');
            const detailModal = document.getElementById('detailModal');
            if (modalClose) {
                modalClose.addEventListener('click', () => {
                    detailModal.classList.remove('active');
                });
                console.log('Modal close event listener added');
            }
            
            if (detailModal) {
                detailModal.addEventListener('click', e => {
                    if (e.target === e.currentTarget) e.currentTarget.classList.remove('active');
                });
                console.log('Modal overlay event listener added');
            }

            // Advisory
            const generateAdvisoryBtn = document.getElementById('generateAdvisoryBtn');
            const approveAdvisoryBtn = document.getElementById('approveAdvisoryBtn');
            
            if (generateAdvisoryBtn) {
                generateAdvisoryBtn.addEventListener('click', generateAdvisory);
                console.log('Generate advisory button event listener added');
            }
            
            if (approveAdvisoryBtn) {
                approveAdvisoryBtn.addEventListener('click', approveAdvisory);
                console.log('Approve advisory button event listener added');
            }

            // Trigger cycle
            const triggerCycleBtn = document.getElementById('triggerCycleBtn');
            if (triggerCycleBtn) {
                triggerCycleBtn.addEventListener('click', triggerCycle);
                console.log('Trigger cycle button event listener added');
            }

            // Check auth
            console.log('Checking authentication status...');
            if (authToken) {
                // Validate token by trying dashboard
                console.log('Validating existing token...');
                apiFetch('/api/gov/dashboard').then(() => {
                    console.log('Token valid, showing app');
                    showApp();
                }).catch((error) => {
                    console.log('Token invalid, showing login', error);
                    showLogin();
                });
            } else {
                console.log('No token found, showing login');
                showLogin();
            }
        } catch (error) {
            console.error('Error during DOM initialization:', error);
            // Fallback: show login screen
            showLogin();
        }
    });

})();

File: backend/routers/government.py
"""
AEGIS Government API — Protected endpoints for Intelligence Command Center.
"""
from typing import Optional
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.orm import Session
from sqlalchemy import desc

from database import get_db
from auth import get_current_user, verify_password, create_access_token
from models import (
    GovernmentUser, RiskAssessment, Alert, Advisory,
    IntelligenceFeedEntry, EnvironmentalReading,
)
from engine.explainer import generate_public_advisory
from engine.intervention_simulator import simulate_intervention
from engine.resource_optimizer import optimize_resources

router = APIRouter(prefix="/api/gov", tags=["Government"])


# ── Request/Response Models ──────────────────────────────

class LoginRequest(BaseModel):
    username: str
    password: str


class AdvisoryRequest(BaseModel):
    message: Optional[str] = None
    severity: str = "Moderate"

class SimulateRequest(BaseModel):
    zone_id: int
    interventions: list[str]

class OptimizeRequest(BaseModel):
    total_budget: float


# ── Auth ─────────────────────────────────────────────────

@router.post("/login")
def login(req: LoginRequest, db: Session = Depends(get_db)):
    user = db.query(GovernmentUser).filter(
        GovernmentUser.username == req.username
    ).first()
    if not user or not verify_password(req.password, user.hashed_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid credentials",
        )
    token = create_access_token(user.username)
    return {"access_token": token, "token_type": "bearer", "username": user.username}


# ── Dashboard ────────────────────────────────────────────

@router.get("/dashboard")
def get_dashboard(
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    latest = (
        db.query(RiskAssessment)
        .order_by(desc(RiskAssessment.timestamp))
        .first()
    )
    if not latest:
        return {
            "air": {"score": 0, "confidence": 0, "trend": "stable", "status": "No Data", "value": 0},
            "heat": {"score": 0, "confidence": 0, "trend": "stable", "status": "No Data", "value": 0},
            "composite": {"score": 0, "confidence": 0, "trend": "stable", "status": "No Data", "value": 0},
            "last_updated": None,
            "active_alerts": 0,
        }

    active_alerts = db.query(Alert).filter(Alert.acknowledged == False).count()

    return {
        "air": {
            "score": latest.air_score,
            "confidence": latest.air_confidence,
            "trend": latest.air_trend,
            "status": _status_label(latest.air_score),
            "value": latest.pm25_value,
            "unit": "µg/m³",
        },
        "heat": {
            "score": latest.heat_score,
            "confidence": latest.heat_confidence,
            "trend": latest.heat_trend,
            "status": _status_label(latest.heat_score),
            "value": latest.heat_index,
            "unit": "°C",
        },
        "composite": {
            "score": latest.composite_score,
            "confidence": latest.composite_confidence,
            "trend": latest.composite_trend,
            "status": _status_label(latest.composite_score),
            "amplification": latest.amplification_factor,
        },
        "last_updated": latest.timestamp.isoformat() if latest.timestamp else None,
        "active_alerts": active_alerts,
    }


# ── Simulation ───────────────────────────────────────────

@router.post("/simulate")
def simulate(
    req: SimulateRequest,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    """Admin endpoint to simulate the effect of interventions."""
    result = simulate_intervention(db, req.zone_id, req.interventions)
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result

@router.post("/optimize-resources")
def optimize_resource_allocation(
    req: OptimizeRequest,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    """Admin endpoint to recommend resource allocation using PuLP."""
    result = optimize_resources(db, req.total_budget)
    if "error" in result:
        raise HTTPException(status_code=500, detail=result["error"])
    return result


# ── Detailed Intelligence ────────────────────────────────

@router.get("/risk/{risk_type}/detail")
def get_risk_detail(
    risk_type: str,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    if risk_type not in ("air", "heat", "composite"):
        raise HTTPException(status_code=400, detail="Invalid risk type")

    latest = (
        db.query(RiskAssessment)
        .order_by(desc(RiskAssessment.timestamp))
        .first()
    )
    if not latest:
        raise HTTPException(status_code=404, detail="No assessment data available")

    base = {
        "timestamp": latest.timestamp.isoformat() if latest.timestamp else None,
        "intelligence_summary": latest.intelligence_summary,
        "anomaly_flags": latest.anomaly_flags,
        "anomaly_description": latest.anomaly_description,
    }

    if risk_type == "air":
        base.update({
            "score": latest.air_score,
            "confidence": latest.air_confidence,
            "trend": latest.air_trend,
            "current_value": latest.pm25_value,
            "unit": "µg/m³",
            "baseline_mean": latest.pm25_baseline_mean,
            "zscore": latest.pm25_zscore,
            "percentile": latest.pm25_percentile,
            "deviation": round(latest.pm25_value - latest.pm25_baseline_mean, 2) if latest.pm25_value and latest.pm25_baseline_mean else None,
        })
    elif risk_type == "heat":
        base.update({
            "score": latest.heat_score,
            "confidence": latest.heat_confidence,
            "trend": latest.heat_trend,
            "current_value": latest.heat_index,
            "unit": "°C",
            "temperature": latest.temperature,
            "humidity": latest.humidity,
            "baseline_mean": latest.heat_baseline_mean,
            "zscore": latest.heat_zscore,
            "percentile": latest.heat_percentile,
            "deviation": round(latest.heat_index - latest.heat_baseline_mean, 2) if latest.heat_index and latest.heat_baseline_mean else None,
        })
    else:  # composite
        base.update({
            "score": latest.composite_score,
            "confidence": latest.composite_confidence,
            "trend": latest.composite_trend,
            "amplification_factor": latest.amplification_factor,
            "air_score": latest.air_score,
            "heat_score": latest.heat_score,
            "air_contribution": round(0.45 * latest.air_score, 1),
            "heat_contribution": round(0.55 * latest.heat_score, 1),
            "synergistic_boost": round(
                (0.45 * latest.air_score + 0.55 * latest.heat_score)
                * (latest.amplification_factor - 1.0), 1
            ) if latest.amplification_factor else 0,
        })

    return base


# ── Alerts ───────────────────────────────────────────────

@router.get("/alerts")
def get_alerts(
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    active = (
        db.query(Alert)
        .filter(Alert.acknowledged == False)
        .order_by(desc(Alert.timestamp))
        .limit(50)
        .all()
    )
    past = (
        db.query(Alert)
        .filter(Alert.acknowledged == True)
        .order_by(desc(Alert.timestamp))
        .limit(50)
        .all()
    )

    def serialize(a):
        return {
            "id": a.id,
            "severity": a.severity,
            "type": a.alert_type,
            "message": a.message,
            "risk_score": a.risk_score,
            "acknowledged": a.acknowledged,
            "timestamp": a.timestamp.isoformat() if a.timestamp else None,
        }

    return {
        "active": [serialize(a) for a in active],
        "past": [serialize(a) for a in past],
    }


@router.post("/alerts/{alert_id}/acknowledge")
def acknowledge_alert(
    alert_id: int,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    alert = db.query(Alert).filter(Alert.id == alert_id).first()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    alert.acknowledged = True
    alert.acknowledged_at = datetime.now(timezone.utc)
    db.commit()
    return {"status": "acknowledged", "id": alert_id}


# ── Advisory ─────────────────────────────────────────────

@router.post("/advisory/generate")
async def generate_advisory(
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    latest = (
        db.query(RiskAssessment)
        .order_by(desc(RiskAssessment.timestamp))
        .first()
    )
    if not latest:
        raise HTTPException(status_code=404, detail="No assessment data")

    risk_data = {
        "air_score": latest.air_score,
        "heat_score": latest.heat_score,
        "composite_score": latest.composite_score,
        "pm25": latest.pm25_value,
        "heat_index": latest.heat_index,
    }

    message = await generate_public_advisory(risk_data)
    severity = _status_label(latest.composite_score)

    advisory = Advisory(message=message, severity=severity)
    db.add(advisory)
    db.commit()
    db.refresh(advisory)

    return {
        "id": advisory.id,
        "message": advisory.message,
        "severity": advisory.severity,
        "approved": False,
    }


@router.post("/advisory/{advisory_id}/approve")
def approve_advisory(
    advisory_id: int,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    advisory = db.query(Advisory).filter(Advisory.id == advisory_id).first()
    if not advisory:
        raise HTTPException(status_code=404, detail="Advisory not found")
    advisory.approved = True
    advisory.approved_at = datetime.now(timezone.utc)
    advisory.published = True
    advisory.published_at = datetime.now(timezone.utc)
    db.commit()

    # Log to feed
    entry = IntelligenceFeedEntry(
        entry_type="advisory",
        title="Public Advisory Published",
        content=advisory.message,
        severity=advisory.severity.lower(),
    )
    db.add(entry)
    db.commit()

    return {"status": "approved_and_published", "id": advisory_id}


# ── Intelligence Feed ────────────────────────────────────

@router.get("/feed")
def get_feed(
    limit: int = 50,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    entries = (
        db.query(IntelligenceFeedEntry)
        .order_by(desc(IntelligenceFeedEntry.timestamp))
        .limit(limit)
        .all()
    )
    return [
        {
            "id": e.id,
            "type": e.entry_type,
            "title": e.title,
            "content": e.content,
            "severity": e.severity,
            "timestamp": e.timestamp.isoformat() if e.timestamp else None,
        }
        for e in entries
    ]


# ── Trends ───────────────────────────────────────────────

@router.get("/trends")
def get_trends(
    hours: int = 24,
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    assessments = (
        db.query(RiskAssessment)
        .order_by(desc(RiskAssessment.timestamp))
        .limit(hours)
        .all()
    )
    assessments.reverse()

    return {
        "labels": [a.timestamp.strftime("%H:%M") if a.timestamp else "" for a in assessments],
        "air_scores": [a.air_score for a in assessments],
        "heat_scores": [a.heat_score for a in assessments],
        "composite_scores": [a.composite_score for a in assessments],
        "pm25_values": [a.pm25_value for a in assessments],
        "heat_index_values": [a.heat_index for a in assessments],
    }


# ── System Health ────────────────────────────────────────

@router.get("/system-health")
def get_system_health(
    db: Session = Depends(get_db),
    user: GovernmentUser = Depends(get_current_user),
):
    latest_reading = (
        db.query(EnvironmentalReading)
        .order_by(desc(EnvironmentalReading.timestamp))
        .first()
    )
    latest_assessment = (
        db.query(RiskAssessment)
        .order_by(desc(RiskAssessment.timestamp))
        .first()
    )

    return {
        "observer_agent": {
            "status": "Active" if latest_reading else "Waiting",
            "last_run": latest_reading.timestamp.isoformat() if latest_reading and latest_reading.timestamp else None,
        },
        "detection_engine": {
            "status": "Running" if latest_assessment else "Idle",
            "last_run": latest_assessment.timestamp.isoformat() if latest_assessment and latest_assessment.timestamp else None,
        },
        "investigation_engine": {"status": "Operational"},
        "explanation_engine": {"status": "Connected"},
        "autonomy_mode": {"status": "Enabled"},
    }


def _status_label(score: float) -> str:
    if score < 40:
        return "Normal"
    elif score < 70:
        return "Moderate"
    return "High"
